### PR TITLE
Port Repair and Migrator events to IEventDispatcher

### LINKED ISSUE #33264: Update via UI breaks with future version of symfony/event-dispatcher
```
TypeError: Symfony\\Component\\EventDispatcher\\GenericEvent::getArgument(): Argument #1 ($key) must be of type string, int given, called in /var/www/nextcloud/server/lib/private/EventDispatcher/GenericEventWrapper.php on line 81 at /var/www/nextcloud/server/appsbabies/spreed/vendor/symfony/event-dispatcher/GenericEvent.php#60
```

We have several events which use int based arrays for the arguments:
- [x] lib/private/DB/Migrator.php
https://github.com/nextcloud/server/blob/215aef3cbdc1963be1bb6bca5218ee0a4b7f1665/lib/private/DB/Migrator.php#L204
- [x] lib/private/Repair.php
https://github.com/nextcloud/server/blob/9a9409e31583c0fdd7c0bfe67f5f9177c7651830/lib/private/Repair.php#L255-L260

We can adjust them already now to have string based arguments which is a much better description.

### PR DESCRIPTION
**Pull request by come-nc**
See https://github.com/nextcloud/server/issues/32128
This also fixes https://github.com/nextcloud/server/issues/33264

### CODE REVIEW

**Comment by come-nc** [comment]
> It seems `\OC\DB\Migrator::checkTable` is actually never dispatched, is it?

Not since a2f3077ee802b4b291ba4f846bfceb69ec0f225f

**Comment by skjnldsv** [comment]
What's going on with the [github-code-scanning](https://github.com/apps/github-code-scanning) bot ?

File: core/ajax/update.php [hunk 1]
```diff
@@ -125,25 +119,19 @@ public function handleRepairFeedback($event) {
...
-			$eventSource->send('success', $l->t('[%d / %d]: %s', [$event[0], $event[1], $event->getSubject()]));
-		}
-	});
-	$dispatcher->addListener('\OC\DB\Migrator::checkTable', function ($event) use ($eventSource, $l) {
-		if ($event instanceof GenericEvent) {
-			$eventSource->send('success', $l->t('[%d / %d]: Checking table %s', [$event[0], $event[1], $event->getSubject()]));
-		}
+	/** @var IEventDispatcher $dispatcher */
+	$dispatcher = \OC::$server->get(IEventDispatcher::class);
+	$dispatcher->addListener(MigratorExecuteSqlEvent::class, function (MigratorExecuteSqlEvent $event) use ($eventSource, $l) {
```
Comment (CarlSchwan) [inline core/ajax/update.php:124]:
```suggestion
	$dispatcher->addListener(MigratorExecuteSqlEvent::class, function (MigratorExecuteSqlEvent $event) use ($eventSource, $l): void {
```

**Review by CarlSchwan — APPROVED** [review]

**Review by CarlSchwan — APPROVED** [review]

File: core/Command/Maintenance/Repair.php [hunk 2]
```diff
@@ -102,47 +109,36 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
+			$this->progress->advance($event->getCurrentStep());
+		} elseif ($event instanceof RepairFinishEvent) {
+			$this->progress->finish();
+			$this->output->writeln('');
+		} elseif ($event instanceof RepairStepEvent) {
+			$this->output->writeln(' - ' . $event->getStepName());
+		} elseif ($event instanceof RepairInfoEvent) {
+			$this->output->writeln('     - ' . $event->getMessage());
+		} elseif ($event instanceof RepairWarningEvent) {
+			$this->output->writeln('     - WARNING: ' . $event->getMessage());
```
Comment (nickvergessen) [inline core/Command/Maintenance/Repair.php:139]: While we are on it:
```suggestion
			$this->output->writeln('<comment>     - WARNING: ' . $event->getMessage() . '</comment>');
```

For yellow font on default background?

File: core/ajax/update.php [hunk 3]
```diff
@@ -48,57 +59,40 @@
...
+			$this->progressStateMax = $this->progressStateStep;
+			$this->eventSource->send('success', $this->l10n->t('[%d / %d]: %s', [$this->progressStateStep, $this->progressStateMax, $this->currentStep]));
+		} elseif ($event instanceof RepairStepEvent) {
+			$this->eventSource->send('success', $this->l10n->t('Repair step:') . ' ' . $event->getStepName());
+		} elseif ($event instanceof RepairInfoEvent) {
+			$this->eventSource->send('success', $this->l10n->t('Repair info:') . ' ' . $event->getMessage());
+		} elseif ($event instanceof RepairWarningEvent) {
+			$this->eventSource->send('notice', $this->l10n->t('Repair warning:') . ' ' . $event->getMessage());
+		} elseif ($event instanceof RepairErrorEvent) {
+			$this->eventSource->send('notice', $this->l10n->t('Repair error:') . ' ' . $event->getMessage());
```
Comment (nickvergessen) [inline core/ajax/update.php:95]: we don't have "error" support there?

File: lib/private/Repair/Events/RepairAdvanceEvent.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
+<?php
```
Comment (nickvergessen) [inline lib/private/Repair/Events/RepairAdvanceEvent.php:1]: declare strict?

File: lib/private/DB/MigratorExecuteSqlEvent.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
+<?php
```
Comment (nickvergessen) [inline lib/private/DB/MigratorExecuteSqlEvent.php:1]: declare strict?

File: lib/private/Repair/Events/RepairAdvanceEvent.php [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program. If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+namespace OC\Repair\Events;
+
+use OCP\EventDispatcher\Event;
+
+class RepairAdvanceEvent extends Event {
+	// TODO Is that current step or step increment?
```
Comment (nickvergessen) [inline lib/private/Repair/Events/RepairAdvanceEvent.php:27]: still open?

File: lib/private/Repair/Events/RepairErrorEvent.php [hunk 7]
```diff
@@ -0,0 +1,38 @@
+<?php
```
Comment (nickvergessen) [inline lib/private/Repair/Events/RepairErrorEvent.php:1]: declare strict?

File: lib/private/Repair/Events/RepairFinishEvent.php [hunk 8]
```diff
@@ -0,0 +1,27 @@
+<?php
```
Comment (nickvergessen) [inline lib/private/Repair/Events/RepairFinishEvent.php:1]: declare strict?

File: lib/private/Repair/Events/RepairInfoEvent.php [hunk 9]
```diff
@@ -0,0 +1,38 @@
+<?php
```
Comment (nickvergessen) [inline lib/private/Repair/Events/RepairInfoEvent.php:1]: declare strict?

File: lib/private/Repair/Events/RepairStartEvent.php [hunk 10]
```diff
@@ -0,0 +1,45 @@
+<?php
```
Comment (nickvergessen) [inline lib/private/Repair/Events/RepairStartEvent.php:1]: declare strict?

File: lib/private/Repair/Events/RepairStepEvent.php [hunk 11]
```diff
@@ -0,0 +1,38 @@
+<?php
```
Comment (nickvergessen) [inline lib/private/Repair/Events/RepairStepEvent.php:1]: declare strict?

File: lib/private/Repair/Events/RepairWarningEvent.php [hunk 12]
```diff
@@ -0,0 +1,38 @@
+<?php
```
Comment (nickvergessen) [inline lib/private/Repair/Events/RepairWarningEvent.php:1]: declare strict?

**Review by nickvergessen — COMMENTED** [review]
Didn't test, but looks good in general

File: core/Command/Maintenance/Repair.php (hunk 2 above)
Comment (come-nc) [reply core/Command/Maintenance/Repair.php:139]: Only for the warning line or all of them?

File: core/ajax/update.php (hunk 3 above)
Comment (come-nc) [reply core/ajax/update.php:95]: No idea, it was like this so I left it.
IEventSource does not say which types are valid.

File: lib/private/Repair/Events/RepairAdvanceEvent.php (hunk 6 above)
Comment (come-nc) [reply lib/private/Repair/Events/RepairAdvanceEvent.php:27]: Yes this is still an open question, I initialy thought it was storing the current step, but the code reacting to the call seem to treat it as an increment, this needs to be cleared out before merging.

File: core/Command/Maintenance/Repair.php (hunk 2 above)
Comment (nickvergessen) [reply core/Command/Maintenance/Repair.php:139]: warning as that is then yellow like the error is then red.

Comment (come-nc) [reply core/Command/Maintenance/Repair.php]: https://github.com/nextcloud/server/pull/33640/commits/652ab6da3f5309b0ea6e1e19ce4d915ecb6cc9b1 ?

File: lib/private/Repair/Events/RepairAdvanceEvent.php (hunk 6 above)
Comment (come-nc) [reply lib/private/Repair/Events/RepairAdvanceEvent.php]: Solved, I checked and this is an increment, pushed a commit to make it clear.

File: core/ajax/update.php (hunk 3 above)
Comment (come-nc) [reply core/ajax/update.php:95]: There is "failure" but it seems only used for non-recoverable errors and is always followed by an exit, which is not the case here.
I do not understand who is consuming this data to check which types are actually understood :-/

File: core/Command/Maintenance/Repair.php (hunk 2 above)
Comment (nickvergessen) [reply core/Command/Maintenance/Repair.php]: We will see if that is too color full 👯

File: core/ajax/update.php (hunk 3 above)
Comment (nickvergessen) [reply core/ajax/update.php:95]: Totally fine then. Was just wondering

Comment (come-nc) [reply core/ajax/update.php]: Carl found the information, there is error level, changed it to error and documented the types in the interface.

**Comment by come-nc** [comment]
Tested web updater with a no-problem update, seemed fine.
Need to test app update, error update and cli update I suppose.

**Comment by come-nc** [comment]
cli upgrade works fine. Not sure how to test app upgrade or buggy upgrade, I test by switching branches so there is no actual download happenning.

**Review by skjnldsv — APPROVED** [review]

**Pull request merged**